// js-brasil/dist/faker.js
"use strict";
/**
 * fakerBr — random valid Brazilian document/data generators
 *
 * This module is intentionally NOT imported by src/index.ts so it is never
 * included in the main validator/masking bundle.  Import it explicitly:
 *
 *   import { fakerBr } from 'js-brasil/faker';
 *   // or (without package exports field):
 *   import { fakerBr } from 'js-brasil/dist/faker';
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.fakerBr = exports.endereco = exports.cor = exports.cidade = exports.estado = exports.email = exports.nomeCompleto = exports.sobrenome = exports.nome = exports.placaMercosul = exports.placa = exports.celular = exports.telefone = exports.cns = exports.certidao = exports.cnh = exports.titulo = exports.pis = exports.renavam = exports.rg = exports.cep = exports.cnpj = exports.cpf = void 0;
var create_1 = require("./create");
var rg_1 = require("./rg");
var utils_1 = require("./utils");
var name_1 = require("./name");
// ── internal helpers ──────────────────────────────────────────────────────────
function rndDigits(n) {
    return Array.from({ length: n }, function () { return Math.floor(Math.random() * 10); }).join('');
}
function rndLetter() {
    return utils_1.randomLetter(1, true);
}
function pad(n, len) {
    var s = String(n);
    while (s.length < len)
        s = '0' + s;
    return s;
}
function pickFrom(arr) {
    return arr[Math.floor(Math.random() * arr.length)];
}
// ── document generators ───────────────────────────────────────────────────────
/** Random valid CPF — 11 digits unformatted */
function cpf() {
    var base = rndDigits(9);
    // avoid all-same-digit sequences (rejected by validator)
    while (/^(\d)\1+$/.test(base)) {
        base = rndDigits(9);
    }
    var digs = create_1.create_cpf(base + '00');
    return base + digs;
}
exports.cpf = cpf;
/** Random valid CNPJ — 14 digits unformatted */
function cnpj() {
    var base = rndDigits(8) + '0001'; // 8 random + branch 0001 = 12 chars
    // d1 depends on the first 12 digits; d2 depends on the first 13 (base + d1)
    var d1 = create_1.create_cnpj(base + '00')[0];
    var d2 = create_1.create_cnpj(base + d1 + '0')[1];
    return base + d1 + d2;
}
exports.cnpj = cnpj;
/** Random CEP — 8 digits unformatted */
function cep() {
    return rndDigits(8);
}
exports.cep = cep;
/** Random valid RG in SP format — 8 digits + check digit (may be 'X') */
function rg() {
    var base = rndDigits(8);
    var dv = rg_1.rg_sp(base);
    return base + dv;
}
exports.rg = rg;
/** Random valid RENAVAM — 11 digits */
function renavam() {
    var base = rndDigits(10) + '0';
    var dv = create_1.create_renavam(base);
    return base.slice(0, 10) + dv;
}
exports.renavam = renavam;
/** Random valid PIS/PASEP — 11 digits */
function pis() {
    var base = rndDigits(10) + '0';
    var dv = create_1.create_pispasep(base);
    return base.slice(0, 10) + dv;
}
exports.pis = pis;
/**
 * Random valid Título Eleitoral — 12 digits
 * Format: [8 sequential][2 state code 01-28][2 check digits]
 */
function titulo() {
    var stateCode = pad(utils_1.randomNumber(1, 29), 2); // '01'..'28'
    var seq = rndDigits(8);
    var base = seq + stateCode + '00'; // 12 chars, placeholder DV
    var dv = create_1.create_titulo_atual(base);
    return seq + stateCode + dv;
}
exports.titulo = titulo;
/** Random CNH number — 11 digits (check digits are the last 2 as-is) */
function cnh() {
    return rndDigits(11);
}
exports.cnh = cnh;
/** Random valid Certidão — 32 digits */
function certidao() {
    var base = rndDigits(30) + '00';
    var dv = create_1.create_certidao(base);
    return base.slice(0, 30) + dv;
}
exports.certidao = certidao;
/**
 * Random valid CNS (Cartão Nacional de Saúde) — 15 digits, provisional format.
 * Starts with 7, 8, or 9 and satisfies the weighted sum % 11 === 0 check.
 */
function cns() {
    while (true) {
        var first = pickFrom(['7', '8', '9']);
        var num = first + rndDigits(14);
        var soma = 0;
        for (var i = 0; i < num.length; i++) {
            soma += parseInt(num[i]) * (15 - i);
        }
        if (soma % 11 === 0) {
            return num;
        }
    }
}
exports.cns = cns;
// ── contact / vehicle ─────────────────────────────────────────────────────────
/** Random Brazilian landline — DDD (11-99) + 8 digits */
function telefone() {
    return pad(utils_1.randomNumber(11, 100), 2) + rndDigits(8);
}
exports.telefone = telefone;
/** Random Brazilian mobile — DDD (11-99) + 9 + 8 digits */
function celular() {
    return pad(utils_1.randomNumber(11, 100), 2) + '9' + rndDigits(8);
}
exports.celular = celular;
/** Random old-format licence plate — AAA0000 */
function placa() {
    return rndLetter() + rndLetter() + rndLetter() + rndDigits(4);
}
exports.placa = placa;
/** Random Mercosul-format licence plate — AAA0A00 */
function placaMercosul() {
    return rndLetter() + rndLetter() + rndLetter() + rndDigits(1) + rndLetter() + rndDigits(2);
}
exports.placaMercosul = placaMercosul;
// ── personal / address ────────────────────────────────────────────────────────
/** Random Brazilian first name */
function nome() {
    var nomes = [
        'Ana', 'Maria', 'Juliana', 'Amanda', 'Fernanda', 'Patrícia', 'Camila', 'Aline', 'Bruna', 'Larissa',
        'João', 'José', 'Antonio', 'Carlos', 'Paulo', 'Pedro', 'Lucas', 'Gabriel', 'Mateus', 'Rafael',
        'Luiz', 'Marcos', 'Rodrigo', 'Eduardo', 'Felipe', 'Diego', 'Bruno', 'Gustavo', 'Leonardo', 'Thiago',
    ];
    return pickFrom(nomes);
}
exports.nome = nome;
/** Random Brazilian surname */
function sobrenome() {
    var sobrenomes = [
        'Silva', 'Santos', 'Oliveira', 'Souza', 'Lima', 'Pereira', 'Ferreira', 'Costa', 'Rodrigues', 'Alves',
        'Nascimento', 'Carvalho', 'Fernandes', 'Gomes', 'Martins', 'Rocha', 'Ribeiro', 'Almeida', 'Mendes', 'Barros',
        'Araújo', 'Cardoso', 'Moreira', 'Nunes', 'Correia', 'Melo', 'Pinto', 'Teixeira', 'Lopes', 'Cruz',
    ];
    return pickFrom(sobrenomes);
}
exports.sobrenome = sobrenome;
/** Random full name (first + last) */
function nomeCompleto() {
    return nome() + ' ' + sobrenome();
}
exports.nomeCompleto = nomeCompleto;
/** Random email based on a generated name */
function email() {
    var domains = ['gmail.com', 'hotmail.com', 'yahoo.com.br', 'outlook.com', 'uol.com.br'];
    var n = nome().toLowerCase().normalize('NFD').replace(/[\u0300-\u036f]/g, '');
    var s = sobrenome().toLowerCase().normalize('NFD').replace(/[\u0300-\u036f]/g, '');
    return n + "." + s + utils_1.randomNumber(1, 99) + "@" + pickFrom(domains);
}
exports.email = email;
/** Random Brazilian state as { nome, uf } */
function estado() {
    return pickFrom(name_1.LOCALIZACAO_ESTADOS);
}
exports.estado = estado;
/** Random Brazilian city name */
function cidade() {
    return pickFrom(name_1.LOCALIZACAO_CIDADES)[0];
}
exports.cidade = cidade;
/** Random Brazilian vehicle colour */
function cor() {
    return pickFrom(utils_1.CORES);
}
exports.cor = cor;
/** Random Brazilian address object */
function endereco() {
    var cidadeEstado = pickFrom(name_1.LOCALIZACAO_CIDADES);
    return {
        logradouro: pickFrom(name_1.LOCALIZACAO_LOGRADOUROS),
        rua: pickFrom(name_1.LOCALIZACAO_RUAS),
        numero: utils_1.randomNumber(1, 9999),
        complemento: pickFrom(name_1.LOCALIZACAO_COMPLEMENTOS),
        bairro: pickFrom(name_1.LOCALIZACAO_BAIRROS),
        cidade: cidadeEstado[0],
        estado: cidadeEstado[1],
        cep: cep(),
    };
}
exports.endereco = endereco;
// ── main export ───────────────────────────────────────────────────────────────
exports.fakerBr = {
    cpf: cpf,
    cnpj: cnpj,
    cep: cep,
    rg: rg,
    renavam: renavam,
    pis: pis,
    titulo: titulo,
    cnh: cnh,
    certidao: certidao,
    cns: cns,
    telefone: telefone,
    celular: celular,
    placa: placa,
    placaMercosul: placaMercosul,
    nome: nome,
    sobrenome: sobrenome,
    nomeCompleto: nomeCompleto,
    email: email,
    estado: estado,
    cidade: cidade,
    cor: cor,
    endereco: endereco,
};
